### ISSUE TITLE
Improve error reporting when RPM resolver fails

### ISSUE THREAD

**Issue by mpas** [suse-edge/edge-image-builder#230]
Sometimes it happens that the RPM resolver fails, the build log shows the exact command that is being executed but the actual error is unclear.

See the example below.. something went wrong but from the error message it is not clear which dependency failed or if there is something else going wrong.

2024-02-24T07:24:43.933Z	FATAL	build/build.go:90	An error occurred building the image: configuring combustion: configuring component "RPM": resolving rpm/package dependencies: building resolver image: building image from context /eib/_build/build-Feb24_07-13-56/resolver-image-build: building at STEP "RUN zypper     --pkg-cache-dir /tmp/rpm-repo     --gpg-auto-import-keys     --no-gpg-checks     install -y     --download-only     --force-resolution     --auto-agree-with-licenses     --allow-vendor-change     -n nano wget2 nvidia-open-driver-G06-signed-kmp-default nvidia-compute-utils-G06 nvidia-container-toolkit nvidia-gl-G06 /tmp/rpm-repo/local/venv-salt-minion-3006.0-150000.3.45.4.x86_64.rpm": while running runtime: exit status 4

Can this error reporting be improved so that it is clear what the actual error is?

**Comment by rdoxenham** [comment]
@mpas please look in the `podman-image-build.log` file, this will give you a more useful output for what happened during the RPM resolver. As an example, on openSUSE Leap Micro, if I try to install some packages that don't exist in the default repo's, the `podman-system-service.log` file shows:

```time="2024-02-15T12:26:47Z" level=debug msg="Error building at step {Env:[PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin] Command:run Args:[zypper     --pkg-cache-dir /tmp/rpm-repo     --gpg-auto-import-keys     install -y     --download-only     --force-resolution     --auto-agree-with-licenses     --allow-vendor-change     -n nano wget2 k3s-selinux] Flags:[] Attrs:map[] Message:RUN zypper     --pkg-cache-dir /tmp/rpm-repo     --gpg-auto-import-keys     install -y     --download-only     --force-resolution     --auto-agree-with-licenses     --allow-vendor-change     -n nano wget2 k3s-selinux Heredocs:[] Original:RUN zypper     --pkg-cache-dir /tmp/rpm-repo     --gpg-auto-import-keys     install -y     --download-only     --force-resolution     --auto-agree-with-licenses     --allow-vendor-change     -n nano wget2 k3s-selinux }: while running runtime: exit status 104"```

Which, as you say, isn't that useful. However, if I look in `podman-image-build.log`:

```
time="2024-02-15T12:26:39Z" level=debug msg="closing stdin"
Retrieving repository 'addrepo0' metadata [.....done]
Building repository 'addrepo0' cache [...done]
Retrieving repository 'Leap Micro Main Repository' metadata [.................done]
Building repository 'Leap Micro Main Repository' cache [....done]
Retrieving repository 'SLE Micro Update Repository' metadata [..............done]
Building repository 'SLE Micro Update Repository' cache [....done]
Loading repository data...
Reading installed packages...
Package 'nano' not found.
Package 'wget2' not found.
time="2024-02-15T12:26:47Z" level=debug msg="\"/var/tmp/buildah3741069997/mnt/rootfs\" is apparently not really mounted, skipping"
time="2024-02-15T12:26:47Z" level=debug msg="\"/var/tmp/buildah3741069997/mnt\" is apparently not really mounted, skipping"
```

Here you can see that `nano` and `wget2` aren't available in openSUSE Leap Micro.

**Comment by mpas** [comment]
@rdoxenham Just checked the package index via (SCC) for SLE Micro and it looks to me that the packages are available in the Suse PackageHub. I will try with the mentioned log files and see what they contain.

**Comment by mpas** [comment]
Just checked the build logs and it seems an update on the https://download.nvidia.com/suse/sle15sp5/x86_64/ repo is causing the issue.

```
Resolving package dependencies...
Problem: nothing provides 'gcc' needed by the to be installed nvidia-driver-G06-kmp-azure-550.54.14_k4.12.14_5.47-0.x86_64
 Solution 1: do not install nvidia-gl-G06-550.54.14-0.x86_64
 Solution 2: break nvidia-driver-G06-kmp-azure-550.54.14_k4.12.14_5.47-0.x86_64 by ignoring some of its dependencies
Choose from above solutions by number or cancel [1/2/c/d/?] (c): c
```

**Comment by atanasdinov** [comment]
@mpas You'll need to manually include a repository as the source of `gcc` in this case.

**Comment by mpas** [comment]
@atanasdinov and @rdoxenham  Thanks for the comments, we indeed are trying to use the ready build drivers. 

The origin of the error that we faced relates to the fact that multiple repositories play an important role. When SLE Micro releases a new driver version on their internal repos, the supporting RPMs that we need are released in the NVIDIA-Suse repository. They are in sync! But we also need stuff from the NVIDIA-Cuda repository which also can contain supporting RPMs that are higher in version then the ones on the NVIDIA-Suse repo. That was the cause of the gcc error!

So for now we fixed the versions of the driver and the supporting RPMs from the NVIDIA-Suse repo that solves a part of the problem. We now need to build an RPM cache of the NVIDIA-Suse repo due to the fact that only new versions are kept over there and older versions discarded. In order to keep our builds working we also need to keep the old RPM's from the NVIDIA-Suse repo.

But again thanks a lot for you comments!!!
